core/src/main/scala/org/apache/spark/launcher/LauncherBackend.scala

```diff
@@ -71,6 +100,9 @@ private[spark] abstract class LauncherBackend {
     if (connection != null && lastState != state) {
       connection.send(new SetState(state))
       lastState = state
+      if (!_isConnected && stopOnShutdown) {
```

Comment:
on the other issue of disconnect() I would think that if we just document that behavior its fine.

I hadn't noticed before that the disconnect java docs state "Disconnects the handle from the application, without stopping it."